buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -197,37 +203,31 @@ private static Queue<Chunk> createSharedChunkQueue() {
     public ByteBuf allocate(int size, int maxCapacity) {
         if (size <= MAX_CHUNK_SIZE) {
             Thread currentThread = Thread.currentThread();
-            boolean willCleanupFastThreadLocals = FastThreadLocalThread.willCleanupFastThreadLocals(currentThread);
-            AdaptiveByteBuf buf = AdaptiveByteBuf.newInstance(willCleanupFastThreadLocals);
-            try {
-                if (allocate(size, maxCapacity, currentThread, buf)) {
-                    // Allocation did work, the ownership will is transferred to the caller.
-                    AdaptiveByteBuf result = buf;
-                    buf = null;
-                    return result;
-                }
-            } finally {
-                if (buf != null) {
-                    // We didn't handover ownership of the buf, release it now so we don't leak.
-                    buf.release();
-                }
+            AdaptiveByteBuf buf = allocate(size, maxCapacity, currentThread, null);
+            if (buf != null) {
+                return buf;
             }
         }
         // The magazines failed us, or the buffer is too big to be pooled.
         return chunkAllocator.allocate(size, maxCapacity);
     }
 
-    private boolean allocate(int size, int maxCapacity, Thread currentThread, AdaptiveByteBuf buf) {
+    private AdaptiveByteBuf allocate(int size, int maxCapacity, Thread currentThread, AdaptiveByteBuf buf) {
```

Comment:
Like stated below I think this method implementation is fragile and confusing. Depending on if buf is null or not we might need to release the returned `AdaptiveByteBuf`. This makes lifecycle management error-prone. Let's rework this somehow

Reply:
I've been thinking a lot about the need to signal failure and the need to return a buffer, and the conflict between those two will always produce complicated code. So instead I'll try to side-step the problem by pushing the fallback paths down, and make this method infallible.